[docs] Start a doc on ClangImporter C->Swift name mapping

jrose-apple opened the pull request:
Includes enums in all their flavors, plus `swift_wrapper` structs.

Planned future content:

- Methods, initializers, properties
  - "Omit-needless-words" rules
  - NSError
  - Boolean properties
  - Inferred default arguments
  - Factory methods as initializers
- Protocols (mainly just that they check for collision)
- CF types (low-priority since non-Apple people can't make them)
- Custom names, including import-as-member and import-as-accessor
- swift_private

jrose-apple commented:
@swift-ci Please smoke test

xymus commented on docs/CImportNameTranslation.md:
````diff
@@ -0,0 +1,191 @@
+}
+```
+
+- Struct name is enum name, optionally dropping the word "Code".
````
Is this optional to Omit-needless-words?

jrose-apple replied:
Whoops. I should probably just say "dropping the suffix `"Code"` if present".

brentdax commented on docs/CImportNameTranslation.md:
```diff
@@ -0,0 +1,191 @@
+This document gives a high-level description of how C and Objective-C declarations are translated to Swift, with particular focus on how names are adjusted. It is not attempting to be a *complete* description of everything the compiler does except with regards to how *names* are transformed; even there, some special cases that only apply to Apple's SDKs have been omitted.
```
This should probably have a title.

jrose-apple replied:
I'm always conflicted about that because the filename is sort of a title, but sure.

beccadax reviewed (changes requested):
This is a very welcome addition to our docs folder!

nkcsgexi reviewed (approved):
Thanks for adding this! Could you mention where the name translation APIs are declared in the compiler source? I presume it's `SwiftNameTranslation.h`?

jrose-apple commented:
swift/AST/SwiftNameTranslation.h is for Swift-to-ObjC, not ObjC-to-Swift. Most of what's talked about here is in swift/ClangImporter/ImportName.h, but not all of it.

jrose-apple commented:
@swift-ci Please smoke test

nkcsgexi commented:
ah, OK. Notice that both directions of translation are interesting for cross-language automatic renaming.

jrose-apple commented:
> Notice that both directions of translation are interesting for cross-language automatic renaming.

I think I'll rename the file to CToSwiftNameTranslation, leaving room for a future SwiftToCNameTranslation.

nkcsgexi commented:
> I think I'll rename the file to CToSwiftNameTranslation, leaving room for a future SwiftToCNameTranslation.

Sounds good, Thank you!

jrose-apple committed:
[docs] Start a doc on ClangImporter C->Swift name mapping

Includes enums in all their flavors, plus swift_wrapper structs.
Planned future content:

- Methods, initializers, properties
  - "Omit-needless-words" rules
  - NSError
  - Boolean properties
  - Inferred default arguments
  - Factory methods as initializers
- Protocols (mainly just that they check for collision)
- CF types (low-priority since non-Apple people can't make them)
- Custom names, including import-as-member and import-as-accessor
- swift_private

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
We've got three approving reviewers so far, so if anyone else has comments we can take it in post-commit review.
